src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -8,16 +8,128 @@
 using System.Collections.Generic;
 using System.Diagnostics;
 using System.Threading;
+using System.Threading.Channels;
 using System.Threading.Tasks;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Diagnostics
 {
+    internal sealed class AsyncQueue<TElement>
+    {
+        private readonly Channel<TElement> _channel = Channel.CreateUnbounded<TElement>();
+        private readonly TaskCompletionSource<bool> _whenCompleted = new(TaskCreationOptions.RunContinuationsAsynchronously);
+
+        /// <summary>
+        /// Gets a task that transitions to a completed state when <see cref="Complete"/> or
+        /// <see cref="TryComplete"/> is called.  This transition will not happen synchronously.
+        /// 
+        /// This Task will not complete until it has completed all existing values returned
+        /// from <see cref="DequeueAsync"/>.
+        /// </summary>
+        public Task WhenCompletedTask
+            => _whenCompleted.Task;
```

Comment:
note: the WhenCompletedTask is highly suspect.  I'm trying to preserve the original semantics, however in practice channel just has someone complete the writechannel, and then the readchannel is done.